### ISSUE TITLE
Async.Wait() faster than Sync calls.

### ISSUE THREAD

**Issue by shachlanAmazon** [StackExchange/StackExchange.Redis#2032]
Our benchmarks have shown that using sync-over-async pattern is consistently faster than using sync methods. 
Getting/setting a string using `db.Wait(db.StringSetAsync("foo", "bar")` performs better in a highly concurrent environment than `db.StringSet("foo", "bar")`. 

A sample benchmark run yielded these results when running against localhost:

```
num threads: 10
sync TPS: 58552 async TPS: 62932

num threads: 20
sync TPS: 59730 async TPS: 106292

num threads: 30
sync TPS: 59780 async TPS: 132813

num threads: 40
sync TPS: 59883 async TPS: 175524

num threads: 50
sync TPS: 58048 async TPS: 161238

num threads: 60
sync TPS: 59408 async TPS: 199906
```

This might mean that there's a low-hanging fruit for users, to simply switch from sync methods to sync-over-async usage.

If you want to repro this, here is [our benchmark program](https://github.com/aws-samples/elasticache-clients-performance-blog-code/blob/mainline/Dotnet/StackExchangeRedis/Program.cs).

**Comment by NickCraver** [comment]
As an example of pipeline usage (without changing your benchmark app too much), this is what I mean:
```cs
private static void AsyncWaitLastWorkerThread(long numCommands)
{
    Task? last = default;
    for (var i = 0; i < numCommands; i++)
    {
        last = getRandomTask();
    }
    if (last is not null)
    {
        db.Wait(last);
    }
}
```
What _should_ be advise is full `async` here, but trying to minimize the comparison. Running all 3 of these now, and only changing the name, you can see the throughput difference more efficiently using tasks and _not_ monopolizing the thread pool (win/win):
```
starting warmup
Completed warmup, duration 34 seconds
Starting SyncWorkerThread Multiplexor Test, Num Threads = 10
Completed Test, TPS: 17403
starting warmup
Completed warmup, duration 39 seconds
Starting AsyncWorkerThread Multiplexor Test, Num Threads = 10
Completed Test, TPS: 17552
starting warmup
Completed warmup, duration 37 seconds
Starting AsyncWaitLastWorkerThread Multiplexor Test, Num Threads = 10
Completed Test, TPS: 117739
starting warmup
Completed warmup, duration 38 seconds
Starting SyncWorkerThread Multiplexor Test, Num Threads = 20
Completed Test, TPS: 30624
starting warmup
Completed warmup, duration 38 seconds
Starting AsyncWorkerThread Multiplexor Test, Num Threads = 20
Completed Test, TPS: 30114
starting warmup
Completed warmup, duration 37 seconds
Starting AsyncWaitLastWorkerThread Multiplexor Test, Num Threads = 20
Completed Test, TPS: 115740
starting warmup
Completed warmup, duration 37 seconds
Starting SyncWorkerThread Multiplexor Test, Num Threads = 30
Completed Test, TPS: 39640
starting warmup
Completed warmup, duration 38 seconds
Starting AsyncWorkerThread Multiplexor Test, Num Threads = 30
Completed Test, TPS: 40074
starting warmup
Completed warmup, duration 38 seconds
Starting AsyncWaitLastWorkerThread Multiplexor Test, Num Threads = 30
Completed Test, TPS: 115830
starting warmup
Completed warmup, duration 38 seconds
Starting SyncWorkerThread Multiplexor Test, Num Threads = 40
Completed Test, TPS: 49319
starting warmup
Completed warmup, duration 38 seconds
Starting AsyncWorkerThread Multiplexor Test, Num Threads = 40
Completed Test, TPS: 50241
starting warmup
Completed warmup, duration 37 seconds
Starting AsyncWaitLastWorkerThread Multiplexor Test, Num Threads = 40
Completed Test, TPS: 113275
starting warmup
Completed warmup, duration 37 seconds
Starting SyncWorkerThread Multiplexor Test, Num Threads = 50
Completed Test, TPS: 56001
starting warmup
Completed warmup, duration 39 seconds
Starting AsyncWorkerThread Multiplexor Test, Num Threads = 50
Completed Test, TPS: 59323
starting warmup
Completed warmup, duration 37 seconds
Starting AsyncWaitLastWorkerThread Multiplexor Test, Num Threads = 50
Completed Test, TPS: 108264
starting warmup
Completed warmup, duration 38 seconds
Starting SyncWorkerThread Multiplexor Test, Num Threads = 60
Completed Test, TPS: 61399
starting warmup
Completed warmup, duration 37 seconds
Starting AsyncWorkerThread Multiplexor Test, Num Threads = 60
Completed Test, TPS: 65992
starting warmup
Completed warmup, duration 38 seconds
Starting AsyncWaitLastWorkerThread Multiplexor Test, Num Threads = 60
Completed Test, TPS: 109809
```

You can ignore the general throughput decline at the end (AV scanner kicked in), but you get the idea: max throughput regardless of thread state. This is the kind of usage we should encourage fo async (and especially pipelined) things :)

**Comment by shachlanAmazon** [comment]
Hi,
Thanks for the quick answer. Yes, I completely agree that proper async/pipelining is the correct way to use the library, but this benchmark was done for a customer who can't refactor to proper async usage right now, so while you're right to add the third option for the sake of future readers and context, it's less relevant to the point I'm trying to make. If I didn't make well, sorry. I'll try to be more explicit :)

> Yes, sync is a tad faster on throughput because it's doing slightly less context switching but at the expense of tying up all of those threads. 

I'm not certain what you meant here. Did you mean "async" instead of sync? Because the benchmark shows that async-over-sync is faster than sync. Does async-over-sync tie up more threads than sync? Does it add additional points of failure?

The reason I'm asking is the sync methods in SE.R are already partially sync-over-async - `PhysicalConnection.FlushSync` waits over `PhysicalConnection.FlushAsync`, and that's called in every sync command AFAIS.
Considering that SE.R is already partially sync-over-async, why not move this higher up in the call stack, and remove a lot of redundant code? It seems to me like there's either an easy performance win to gain from converting sync commands to be sync-over-async internally, or (more likely) there's some amount of flab in the sync flow, which adds overhead above the sync-over-async alternative, and can be removed.

I'd be happy to assist with an investigation, if that will help.

**Comment by NickCraver** [comment]
> I'm not certain what you meant here. Did you mean "async" instead of sync? Because the benchmark shows that async-over-sync is faster than sync. Does async-over-sync tie up more threads than sync? Does it add additional points of failure?

Both are sync, sorry for poor wording there. But the further out you wait from the async wrapped portion, the more thread time you're monopolizing. The benchmark doesn't make throughput for actual workloads faster. It makes throughput _for that benchmark_ slightly faster _at huge threadpool expense_. There's a huge tradeoff, and for almost all users of actual applications doing work (_especially_ if any of it depends on async), it will be a net worse outcome. We see this issue often filed here in one form or another. We absolutely do not want to encourage it - people very, very often do not understand the consequences of thread exhaustion.

> The reason I'm asking is the sync methods in SE.R are already partially sync-over-async - `PhysicalConnection.FlushSync` waits over `PhysicalConnection.FlushAsync`, and that's called in every sync command AFAIS.

Yes, you're correct that there's sync-over-async at the pipe level for lack of _any other option_ -it's correctly a fully async API. You'll also note that all the sync methods on top of it are marked as Obsolete :)

I'm not sure the best way to stress the overall: making your benchmark faster is not a real-world application and monopolizes the entire thread pool. Any app doing actual async work is going to need those threads for other things and it's far better to free them up. As illustrated in my test, I'm running lots of async IO and waiting once - it's not ideal but it's far better than synchronously blocking on _every_ operation and occupying a thread the entire duration. It's still sync on the outside - I don't recommend it, but it's a middle ground.

Also note that in my results they're _drastically_ different from your initial test (this will vary by core count), to where performance is within jitter until 50+ concurrent threads. That's because this machine is 16 physical/32 logical cores and when you exceed that then you'll see a bigger diff. Still, note that async batching is tremendously faster in _all_ cases.

In short: there are a lot of variables here, but regardless of combination: in none of them would I recommend sync-over-async :)

**Comment by nihohit** [comment]
I hope that I'm not exhausting your good will here, but I do really want to understand this issue well. 

> But the further out you wait from the async wrapped portion, the more thread time you're monopolizing.

This seems counter intuitive to me. Just to make sure that I understand this properly, let's say that we have two functions:

```cs
void foo1() {
  Task thingToDo = DoSomethingAsync();
  thingToDo.Wait();
  DoSomethingElse();
}

void foo2() {
  Task thingToDo = DoSomethingAsync();
  DoSomethingElse();
  thingToDo.Wait();
}
```

You say that `foo1` us better than `foo2`, because it waits closer to the created task? Because to me `foo2` seems like the better option - it allows `DoSomethingAsync` and `DoSomethingElse` to run concurrently, which means that in a low-thread usage app you get a chance to increase your concurrency, and in a high-thread usage app you're not losing anything - at both functions, the thread that runs `DoSomethingAsync` will take the same time to run, and the thread that runs `fooX` will take the time that `DoSomethingAsync` and `DoSomethingElse` take to run. Am I missing something?

**Comment by NickCraver** [comment]
@nihohit You're comparing something else - a sequence of commands. We're talking about a _stack_ of commands here, and how much of _an entire stack_ is blocking a thread for. The wider the block, the more the monopolization of resources. "Everything sync" is at one end of the spectrum, "everything async" is at the other.

**Comment by nihohit** [comment]
Sorry, it's @shachlanAmazon, I just I moved to my home computer :).

I'm not sure how a stack of commands is different from a sequence, for this matter. After all, a stack of commands is flattened by the compiler to a sequence of commands, hoepfully without adding too many `jmp` instructions in between.

 AFAIS here `FooSyncOverAsync`, just like `foo2` before, is the better option.
 
```cs
Task BazAsync() {
  return DoSomethingAsync();
}

void BazSync() {
  DoSomethingAsync().Wait();
}

Task BarAsync() {
  var task = DoSomethingAsync();
  DoSomethingElse();
  return task;
}

void BarSync() {
  BazSync();
  DoSomethingElse();
}

void FooSync() {
  BarSync();
}

void FooSyncOverAsync() {
  BarAsync.Wait();
}
```

and 

> "Everything sync" is at one end of the spectrum, "everything async" is at the other

But sync-over-async with more time between starting async tasks and waiting for them seems closer to "everything async" than sync is.

**Comment by NickCraver** [comment]
@nihohit Because in async/await things can be handed off to other threads (granted, there's some context overhead but your wins are: more available threads and usable concurrency). Let's be really clear here: _none_ of these are good options. You want to go `async` - spending a lot of time supporting the in-between state isn't something we want to do, we just don't have the time (and this is all personal time). If you want to go full sync: yes, good luck! Our advice is: don't do that.

**Comment by nihohit** [comment]
Again, `async/await` is better, and sync methods surrounding inherently async operations are bad. I agree. It's just not a scenario my client can support without a major refactor right now, since his app is already fully sync. The only question is why not to implement the the top level sync methods by just waiting over async methods, and thus removing a lot of sync code, while potentially also getting a slight perf improvement over the current sync implementation. It seems like a win-win - internal obsolete code gets deleted, and users get a slight perf win.

But I get the feeling that this discussion has run its course. Thanks for your time :).
